python/pylib/pose.py

```diff
@@ -20,912 +16,1317 @@ def quatInit():
+    DCM[:,1,0] = 2.0 * (q[:,1]*q[:,2] - q[:,0]*q[:,3])
+    DCM[:,1,1] = 1.0 - 2.0 * (q[:,1]**2 + q[:,3]**2)
+    DCM[:,1,2] = 2.0 * (q[:,2]*q[:,3] + q[:,0]*q[:,1])
+    DCM[:,2,0] = 2.0 * (q[:,1]*q[:,3] + q[:,0]*q[:,2])
+    DCM[:,2,1] = 2.0 * (q[:,2]*q[:,3] - q[:,0]*q[:,1])
+    DCM[:,2,2] = 1.0 - 2.0 * (q[:,1]**2 + q[:,2]**2)
+
+    if array == 0:
+        DCM = np.squeeze(DCM)
     return DCM
 
-#  * This will construct quaternions from a direction cosine 
-#  * matrix in the standard rotation sequence.
-#  * [phi][theta][psi] from NED to body frame
-#  *
-#  * body = tBL(3,3)*NED
-#  * q(4,1)
-#  *
-#  * Reference: http://en.wikipedia.org/wiki/Rotation_representation_%28mathematics%29
-def DCMquat(mat):
-    q = np.zeros(4)
-
-    q[0] = 0.5 * np.sqrt(1.0 + mat[0,0] + mat[1,1] + mat[2,2])
-    
-    d = 1.0 / (4.0 * q[0])
-    
-    q[1] = d * (mat[1,2] - mat[2,1])
-    q[2] = d * (mat[2,0] - mat[0,2])
-    q[3] = d * (mat[0,1] - mat[1,0])
 
+# This will construct quaternions from a direction cosine
+# matrix in the standard rotation sequence.
+# [phi][theta][psi] from reference to body frame
+#
+# body = tBL(3,3)*NED
+#
+# Reference: http://en.wikipedia.org/wiki/Rotation_representation_%28mathematics%29
+def DCMquat(mat):
+    if len(np.shape(mat)) == 2:
+        mat = np.expand_dims(mat, axis=0)
+        array = 0
+    else:
+        array = 1
+
+    q = np.empty(shape=(np.shape(mat)[0], 4))
+    q_sq4 = np.empty_like(q)
+
+    q_sq4[:,0] = 1.0 + mat[:,0,0] + mat[:,1,1] + mat[:,2,2]
+    q_sq4[:,1] = 1.0 + mat[:,0,0] - mat[:,1,1] - mat[:,2,2]
+    q_sq4[:,2] = 1.0 - mat[:,0,0] + mat[:,1,1] - mat[:,2,2]
+    q_sq4[:,3] = 1.0 - mat[:,0,0] - mat[:,1,1] + mat[:,2,2]
+
+    for i in range(0, np.shape(mat)[0]):
+        ind = np.argmax(q_sq4[i, :])
+        q[i,ind] = 0.5 * np.sqrt(q_sq4[i,ind])
+        d = 0.25 / q[i,ind]
+        if ind == 0:
+            q[i,1] = d * (mat[i,1,2] - mat[i,2,1])
+            q[i,2] = d * (mat[i,2,0] - mat[i,0,2])
+            q[i,3] = d * (mat[i,0,1] - mat[i,1,0])
+        elif ind == 1:
+            q[i,0] = d * (mat[i,1,2] - mat[i,2,1])
+            q[i,2] = d * (mat[i,1,0] + mat[i,0,1])
+            q[i,3] = d * (mat[i,0,2] + mat[i,2,0])
+        elif ind == 2:
+            q[i,0] = d * (mat[i,2,0] - mat[i,0,2])
+            q[i,1] = d * (mat[i,1,0] + mat[i,0,1])
+            q[i,3] = d * (mat[i,2,1] + mat[i,1,2])
+        else:
+            q[i,0] = d * (mat[i,0,1] - mat[i,1,0])
+            q[i,1] = d * (mat[i,0,2] + mat[i,2,0])
+            q[i,2] = d * (mat[i,2,1] + mat[i,1,2])
+
+    ind = q[:,0] < 0
+    q[ind, :] = -q[ind, :]
+
+    if array == 0:
+        q = np.squeeze(q)
     return q
 
 
-#  * This will construct the euler omega-cross matrix
-#  * wx(3,3)
-#  * p, q, r (rad/sec)
-def eulerWx( euler ):
-    mat = np.zeros((3,3))
+# This will construct the cross-product matrix Wx(3,3)
+# such that cross(x, y) = Wx * y
+# p, q, r (rad/sec)
+def eulerWx(x):
```

Comment:
The name is not very descriptive. Maybe something like crossMatrix() would be better?